store/storethehash/storethehash.go

```diff
@@ -66,14 +70,66 @@ func (s *sthStorage) get(k []byte) ([]indexer.Value, bool, error) {
 		return nil, false, nil
 	}
 
-	out, err := indexer.Unmarshal(value)
+	out, err := indexer.UnmarshalValues(value)
 	if err != nil {
 		return nil, false, err
 	}
 	return out, true, nil
 
 }
 
+func (s *sthStorage) ForEach(iterFunc indexer.IterFunc) error {
+	// It is necessary to do both flush and sync before creating an iterator to
+	// read values that have not been written to file yet
+	s.primary.Flush()
+	err := s.primary.Sync()
+	if err != nil {
+		return err
+	}
+	iter, err := s.primary.Iter()
+	if err != nil {
+		return err
+	}
+
+	// Create a list of unique keys, and then call get for each unique key.
+	//
+	// This is necessary because the primary iterator returns all versions of
+	// each index.  If a key has been updated with new values, the iterator
+	// returns the key with the original values, then again with the values for
+	// the next update, and so on for as many updates a there were for a key.
+	//
+	// TODO: if there are many keys, this could take up too much memory.  Need
```

Comment:
Instead of building the iterator by going through the primary storage (which in the current implementation without compaction may translate into encountering several outdated keys several times as you clearly identified), we could maybe expose `scanIndex` (https://github.com/ipld/go-storethehash/blob/d72ca9b0968c52b249dc13e901757f1aafcbd5eb/store/index/index.go#L154, and iterate only the up to date keys in the index.

Reply:
I looked at at that, but it seems that would require keeping all the keys **and values** in memory.  Maybe I need to look at it again.

Reply:
We can maybe expose an `IndexIterator` so we can process as we receive buckets: https://github.com/ipld/go-storethehash/blob/d72ca9b0968c52b249dc13e901757f1aafcbd5eb/store/index/index.go#L197

In any case, the index is just a few hundred MB at most, I think we can afford keeping it in memory. As we iterate on the keys in each bucket, we get the value from primary and process it, I don't think we need to keep the values also in memory. Does this make sense?

Reply:
In any case, an efficient implementation of this is probably not a priority right now (especially if initially it just going to be used for debugging purposes), so if we see this is going to be too complex, let's go with your current approach, document this in an issue, and focus on the MVP.